### ISSUE TITLE
vpncloud is not working properly in some complicate NAT scenario.

### ISSUE THREAD

**Issue by yuko1225** [dswd/vpncloud.rs#13]
Hello,

It's quite long but please be patient.First...Just look at this network
![20160827144549](https://cloud.githubusercontent.com/assets/6983761/18025741/cbe3f9e2-6c65-11e6-8fe7-fe37f56b3d9e.png)
Node1 is behind NAT and R1 has 2 WAN IP.
The default route of R1 is eth1.
111.111.122.2 is routed to eth0.
Node3 DID HAS a public address.

All peer is configured in tun + switched mode.Peer list of Node1:
```
peers:
  - 111.100.100.100(Node2):9000
  - 112.99.99.99(Node4):9000
  - 111.111.122.2(Node3):9000
```
Peer list of Node3:
```
peers:
  - 111.100.100.100(Node2):9000
  - 112.99.99.99(Node3):9000
```
Node2 and Node 4 have an empty list.

I start Node2 and Node4 first,then I start Node3.Everything is fine right now.Until I bring up Node1....Boom

On Node2 and Node4 ping Node1 I got:
```
PING 10.0.0.1 (10.0.0.1) 56(84) bytes of data.
64 bytes from 10.0.0.1: icmp_req=1 ttl=63 time=170 ms
From 10.0.0.3: icmp_seq=2 Redirect Host(New nexthop: 10.0.0.1)
64 bytes from 10.0.0.1: icmp_req=2 ttl=63 time=165 ms
From 10.0.0.3: icmp_seq=3 Redirect Host(New nexthop: 10.0.0.1)
64 bytes from 10.0.0.1: icmp_req=3 ttl=63 time=255 ms
From 10.0.0.3: icmp_seq=4 Redirect Host(New nexthop: 10.0.0.1)
64 bytes from 10.0.0.1: icmp_req=4 ttl=63 time=163 ms
```
Node1 ping Node4:
```
PING 10.0.0.4 (10.0.0.4) 56(84) bytes of data.
64 bytes from 10.0.0.4: icmp_req=1 ttl=64 time=149 ms
64 bytes from 10.0.0.4: icmp_req=1 ttl=63 time=227 ms (DUP!)
From 10.0.0.3: icmp_seq=2 Redirect Host(New nexthop: 10.0.0.4)
64 bytes from 10.0.0.4: icmp_req=2 ttl=63 time=231 ms
From 10.0.0.3: icmp_seq=3 Redirect Host(New nexthop: 10.0.0.4)
64 bytes from 10.0.0.4: icmp_req=3 ttl=63 time=226 ms
From 10.0.0.3: icmp_seq=4 Redirect Host(New nexthop: 10.0.0.4)
64 bytes from 10.0.0.4: icmp_req=4 ttl=63 time=232 ms
From 10.0.0.3: icmp_seq=5 Redirect Host(New nexthop: 10.0.0.4)
64 bytes from 10.0.0.4: icmp_req=5 ttl=63 time=226 ms
```
Node1 ping Node3
```
PING 10.0.0.2 (10.0.0.2) 56(84) bytes of data.
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=32.1 ms
64 bytes from 10.0.0.2: icmp_req=1 ttl=63 time=166 ms (DUP!)
From 10.0.0.3: icmp_seq=2 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=2 ttl=63 time=166 ms
From 10.0.0.3: icmp_seq=3 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=3 ttl=63 time=174 ms
```
Even from node4 to node2.I got issue
```
PING 10.0.0.2 (10.0.0.2) 56(84) bytes of data.
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=116 ms
64 bytes from 10.0.0.2: icmp_req=1 ttl=63 time=194 ms (DUP!)
From 10.0.0.3: icmp_seq=2 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=2 ttl=63 time=199 ms
From 10.0.0.3: icmp_seq=3 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=3 ttl=63 time=194 ms
^C
--- 10.0.0.2 ping statistics ---
3 packets transmitted, 3 received, +1 duplicates, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 116.744/176.321/199.377/34.454 ms
```
And latency is higher then normal.

I think vpncloud is get confused by the peer address it learned.But you know,this scenario is quite common in what so called CGN(Carrier-grade NAT) which ISP will only assign client a private address.In CGN,ISP use a "public address pool" and every connections(based on connection session!!not private IP!!) will be load balanced to this pool,which means every connections will probably NAT to different public address.

If I'm not guess wrong.There is some suggestion that might help to solve this issue.
1.Make the manual configured peer list to the highest priority.If the peer is already have an IP address in manual configured peer list,ignore other address that learned from other peers.
2.Incoming connection should be use in the first priority if that peer is not in manual peer list.
3.If a peer(learned from other peer) has more than one IP address,not in the manual peer list and no incoming connection.Try all address and select the one that works.

And also,try to let peers to determine each other is behind NAT or not.If both peer is behind NAT.Try NAT punch.If failed,the one has public address should relay the packet transparently.

BTW,Even I strip out the DNAT proxy I still got same issue if I route Node3 to eth0 on R1

**Comment by dswd** [comment]
Ok, honestly I had a hard time understanding your scenario. But I think I got it now.

Your setup is very complicated and the fact that the same node has different IPs depending on who it talks to is something I did not think about when designing VpnCloud. Considering that, it is great to see it working at all under these conditions.

However, there are two obvious issues here: increased latency and duplicate ping replies. Increased latency is very curious as it usually means that packets take a different path.

I think one of your nodes has IP forwarding activated. If that is right, this is what I think happens:
- When you send the first packet to an unknown address (e.g. from Node1 to Node4), it is broadcast to all peer as you are using switch mode.
- It is received directly by Node4 and it sends a reply. Also the source address is learnt correctly.
- A wrong peer (e.g. Node3) also receives it but instead of just dropping it, it forwards it (as instructed by IP forwarding) to Node4.
- Node4 receives the wrongly forwarded packet. An here the error happens: it learns that the VPN IP of Node1 is associated with Node3. So it will send the reply back to Node3 which will forward it to Node1.
- Then Node1 receives the reply and again the error occurrs: It associates the VPN IP of Node4 with Node3.
- After that Node1 and Node4 know the forwarding and will not broadcast anymore.

The problem is that the indirect path always takes longer and therefore it overrides the correct entries in the switch table. I am not sure if I can change that behavior without breaking some features.
To solve the problem for you, you could either:
- Disable IP forwarding on all nodes
- Use TAP devices instead (as they do not forward)
- Use routed mode instead (so that no learning takes place)

I hope that my analysis is correct. Please check the output of tracepath. If I am correct, you should see another node as intermediate hop.

**Comment by yuko1225** [comment]
@dswd Thanks a lot.I indead saw Node3 forward the packet.But I have to enable the forwarding.I will test the TAP and get back to you shortly.

**Comment by yuko1225** [comment]
@dswd After I set to TAP mode.It seem better(no more Nexthop and no more DUP) but still have some abnormal latency.
Node1 ping Node3
 ```
PING 10.0.0.3 (10.0.0.3) 56(84) bytes of data.
64 bytes from 10.0.0.3: icmp_req=1 ttl=64 time=37.1 ms
64 bytes from 10.0.0.3: icmp_req=2 ttl=64 time=53.3 ms
64 bytes from 10.0.0.3: icmp_req=3 ttl=64 time=76.4 ms
64 bytes from 10.0.0.3: icmp_req=4 ttl=64 time=52.3 ms
64 bytes from 10.0.0.3: icmp_req=5 ttl=64 time=110 ms
64 bytes from 10.0.0.3: icmp_req=6 ttl=64 time=36.9 ms
64 bytes from 10.0.0.3: icmp_req=7 ttl=64 time=37.8 ms
64 bytes from 10.0.0.3: icmp_req=8 ttl=64 time=50.2 ms
64 bytes from 10.0.0.3: icmp_req=9 ttl=64 time=37.3 ms
64 bytes from 10.0.0.3: icmp_req=10 ttl=64 time=37.6 ms
64 bytes from 10.0.0.3: icmp_req=11 ttl=64 time=110 ms
64 bytes from 10.0.0.3: icmp_req=12 ttl=64 time=37.4 ms
64 bytes from 10.0.0.3: icmp_req=13 ttl=64 time=145 ms
64 bytes from 10.0.0.3: icmp_req=14 ttl=64 time=49.9 ms
64 bytes from 10.0.0.3: icmp_req=15 ttl=64 time=153 ms
64 bytes from 10.0.0.3: icmp_req=16 ttl=64 time=49.8 ms
64 bytes from 10.0.0.3: icmp_req=17 ttl=64 time=41.9 ms
64 bytes from 10.0.0.3: icmp_req=18 ttl=64 time=49.9 ms
64 bytes from 10.0.0.3: icmp_req=19 ttl=64 time=76.7 ms
64 bytes from 10.0.0.3: icmp_req=20 ttl=64 time=50.8 ms
64 bytes from 10.0.0.3: icmp_req=21 ttl=64 time=37.1 ms
64 bytes from 10.0.0.3: icmp_req=22 ttl=64 time=37.0 ms
```
Node2 ping Node1
```
64 bytes from 10.0.0.1: icmp_req=1 ttl=64 time=32.6 ms
64 bytes from 10.0.0.1: icmp_req=2 ttl=64 time=81.5 ms
64 bytes from 10.0.0.1: icmp_req=3 ttl=64 time=31.7 ms
64 bytes from 10.0.0.1: icmp_req=4 ttl=64 time=81.0 ms
```

The latency is unstable with some pattern and I'm sure this is not caused by QoS or ISP(Because I have another VPN protocol on UDP and they runs pretty good).Tcpdump on the vpn interface did not reveal any useful information.I did see the ARP packets and it cause the first packet get high RTT but it doesn't make sense to other packets.

Only Node2 ping Node4 didn't get any issue and they're all have empty node list.I tried to route all traffice to eth1 on R1 and strip out DNAT but it doesn't help.

**Comment by dswd** [comment]
Ok that looks better.
I just checked my own vpns and on some I also see this latency pattern. I have to debug it and check where that comes from.

**Comment by dswd** [comment]
Ok, I fixed the latency issue. The problem was that under some conditions, it would resolve peer addresses every second which could delay packet processing by a few milliseconds.

**Comment by yuko1225** [comment]
@dswd That's great.I'm now complie new version and will give you feedback shortly.But I still hope you can solve the CGN problem.Actually it's very common in mobile environment.

**Comment by yuko1225** [comment]
@dswd Thanks a lot.The latency issue is solved.

**Comment by dswd** [comment]
Please check the latest commit. It could potentially avoid the forwarding problem.

**Comment by yuko1225** [comment]
@dswd The first packet I got A LOT OF DUPS.But after that eveything seem fine.I think maybe it still has something that need to figure out.
```
PING 10.0.0.2 (10.0.0.2) 56(84) bytes of data.
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=32.9 ms
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=67.6 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=148 ms (DUP!)
From 10.0.0.4: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=154 ms (DUP!)
From 10.0.0.3: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=188 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=194 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=275 ms (DUP!)
From 10.0.0.4: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=280 ms (DUP!)
From 10.0.0.3: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=315 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=320 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=401 ms (DUP!)
From 10.0.0.4: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=406 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=441 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=446 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=528 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=532 ms (DUP!)
From 10.0.0.3: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=567 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=572 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=654 ms (DUP!)
From 10.0.0.4: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=659 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=694 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=699 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=781 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=785 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=820 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=826 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=908 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=915 ms (DUP!)
From 10.0.0.3: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=947 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=952 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=2 ttl=64 time=32.5 ms
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1033 ms (DUP!)
From 10.0.0.4: icmp_seq=1 Redirect Host(New nexthop: 10.0.0.2)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1039 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1073 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1078 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1160 ms (DUP!)
64 bytes from 10.0.0.2: icmp_req=1 ttl=64 time=1165 ms (DUP!)
```

**Comment by dswd** [comment]
Those duplicates are to be expected when you combine a broadcasting method (learning switch) with ip forwarding. There is not much VpnCloud can do about that.

**Comment by yuko1225** [comment]
@dswd I found another issue with NAT.After a while(I'm not sure how long) with no traffic IN or OUT in Node1.Other nodes couldn't ping Node1.Only after Node1 ping other nodes then others Nodes can ping back.
